### PR TITLE
[WebsocketExchange] Fix no order producer running when using WS

### PR DESCRIPTION
**Pull request by Herklos**
Depends on https://github.com/Drakkar-Software/OctoBot-Trading/pull/344

### CODE REVIEW

File: octobot_trading/channels/orders.py [hunk 1]
```diff
@@ -113,11 +113,14 @@ async def update_order_from_exchange(self, order,
...
-                                       force_job_execution=force_job_execution,
-                                       wait_for_refresh=wait_for_refresh)
+        try:
+            await get_chan(ORDERS_CHANNEL, self.channel.exchange_manager.id).producers[-1]. \
+                update_order_from_exchange(order=order,
+                                           should_notify=should_notify,
+                                           force_job_execution=force_job_execution,
+                                           wait_for_refresh=wait_for_refresh)
+        except IndexError:
+            self.logger.warning("Can't synchronize order from exchange")
```
Comment (GuillaumeDSM) [inline octobot_trading/channels/orders.py:123]: Should this be an error ? 
I think if we really need to update an order (if a websocket didn't receive an order update for ex), we should be able to update it via the rest interface.

Comment (Herklos) [reply octobot_trading/channels/orders.py:123]: I don't know how we can do that with the current architecture

File: octobot_trading/channels/orders.py [hunk 2]
```diff
@@ -104,20 +105,34 @@ async def handle_post_open_order_update(self, symbol, orders, has_new_order):
...
+        try:
+            await (get_chan(ORDERS_CHANNEL, self.channel.exchange_manager.id).producers[-1].
+                   update_order_from_exchange(order=order,
+                                              should_notify=should_notify,
+                                              force_job_execution=force_job_execution,
+                                              wait_for_refresh=wait_for_refresh))
+        except IndexError:
+            if not self.channel.exchange_manager.is_simulated and should_create_try_to_create_producer:
+                self.logger.debug("Missing orders producer, starting one...")
+                await create_authenticated_producer_from_parent(self.channel.exchange_manager,
```
Comment (GuillaumeDSM) [inline octobot_trading/channels/orders.py:128]: 👍

File: octobot_trading/channels/orders.py [hunk 3]
```diff
@@ -104,20 +105,34 @@ async def handle_post_open_order_update(self, symbol, orders, has_new_order):
     async def update_order_from_exchange(self, order,
                                          should_notify=False,
                                          wait_for_refresh=False,
-                                         force_job_execution=False):
+                                         force_job_execution=False,
+                                         should_create_try_to_create_producer=True):
```
Comment (GuillaumeDSM) [inline octobot_trading/channels/orders.py:109]: I think naming this param something like `create_order_producer_if_missing` would be better

File: octobot_trading/exchanges/exchange_channels.py [hunk 4]
```diff
@@ -42,29 +44,50 @@ async def create_exchange_producers(exchange_manager) -> None:
...
-            or exchange_manager.is_simulated
-            or exchange_manager.is_backtesting) \
-            and exchange_manager.trader and exchange_manager.is_trading \
-            and not exchange_manager.is_collecting:
+    if _should_create_simulated_producers(exchange_manager):
         for updater in get_authenticated_updater_simulator_producers():
             await updater(get_chan(updater.CHANNEL_NAME, exchange_manager.id)).run()
 
 
+def _should_create_authenticated_producers(exchange_manager):
```
Comment (GuillaumeDSM) [inline octobot_trading/exchanges/exchange_channels.py:62]: 👍

File: octobot_trading/exchanges/exchange_channels.py [hunk 5]
```diff
@@ -42,29 +44,50 @@ async def create_exchange_producers(exchange_manager) -> None:
...
+    :param exchange_manager: the related exchange manager
+    :return: True if should create simulated producers
+    """
+    return (not exchange_manager.exchange.is_authenticated
+            or exchange_manager.is_simulated
+            or exchange_manager.is_backtesting) \
+           and exchange_manager.trader and exchange_manager.is_trading and not exchange_manager.is_collecting
+
+
+def _should_create_unauthenticated_producers(exchange_manager):
```
Comment (GuillaumeDSM) [inline octobot_trading/exchanges/exchange_channels.py:83]: 👍

File: octobot_trading/exchanges/exchange_channels.py [hunk 6]
```diff
@@ -42,29 +44,50 @@ async def create_exchange_producers(exchange_manager) -> None:
...
+    """
+    :param exchange_manager: the related exchange manager
+    :return: True if should create authenticated producers
+    """
+    return exchange_manager.exchange.is_authenticated \
+           and exchange_manager.trader and exchange_manager.is_trading \
+           and not (exchange_manager.is_simulated or exchange_manager.is_backtesting or exchange_manager.is_collecting)
+
+
+def _should_create_simulated_producers(exchange_manager):
```
Comment (GuillaumeDSM) [inline octobot_trading/exchanges/exchange_channels.py:72]: 👍

File: octobot_trading/exchanges/exchange_channels.py [hunk 7]
```diff
@@ -85,6 +108,56 @@ async def _create_authenticated_producers(exchange_manager) -> None:
...
+                exchange_manager.logger.exception(e, True,
+                                                  f"Error when initializing data for {producer.CHANNEL_NAME} "
+                                                  f"channel required by websocket: {e}")
+    else:
+        # no websocket for this channel: start an producer
+        await producer_instance.run()
+    return producer_instance
+
+
+async def create_authenticated_producer_from_parent(exchange_manager,
```
Comment (GuillaumeDSM) [inline octobot_trading/exchanges/exchange_channels.py:134]: 💯

File: octobot_trading/channels/orders.py (hunk 3 above)
Comment (Herklos) [reply octobot_trading/channels/orders.py:109]: Yes good idea :+1:

**Review by GuillaumeDSM — APPROVED** [review]

**Pull request merged**